tests/suites/0_stateless/12_time_travel/12_0004_time_travel_select_at.sh

```diff
@@ -8,6 +8,10 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 echo "create table t12_0004(c int)" | $MYSQL_CLIENT_CONNECT
 echo "two insertions"
 echo "insert into t12_0004 values(1),(2)" | $MYSQL_CLIENT_CONNECT
+
+# Ensure there is a small interval between two insertions, for time travel
+sleep 0.001
```

Comment:
will this test case fail without sleeping? the timestamp of the table snapshot is supposed to be monotonically increased even if the local clock drifting presents

Reply:
> will this test case fail without sleeping? the timestamp of the table snapshot is supposed to be monotonically increased even if the local clock drifting presents

Generally not, it only fails in the interval of two insertions less than 1 microsecond.

I found `snapshot.timestamp` precision on Linux is nanoseconds(on MacOS is microseconds). But in the sql statement, I can only input/output timestamp with microsecond precision. 

For example, the timestamp of two insertion snapshots is t1(`2021-08-04T13:57:08.427902838`) and t2 (`2021-08-04T13:57:12.427902838`).
I want to find a time-point between t1 and t2, to test time travel only contains snapshot1.

But the result of `select timestamp from fuse_snapshot($first_insertion)` will be `2021-08-04T13:57:08.427902`. If I pass it to `select … from … (AT TIMESTAMP)`, [FuseTable.navigate_to_time_point](https://github.com/datafuselabs/databend/blob/b6ed4596db4f6cd76617996ccc9588f083d69016/query/src/storages/fuse/operations/navigate.rs#L39) will be false. 
So I add 1 microsecond at [33L](https://github.com/datafuselabs/databend/pull/6056/files#diff-560b9c55798cd4ca36b1748bbc30eef560e9885cdabd30da058560b4b344f0baR33), to make sure time-point is larger than t1. But if t2 - t1 < 1 microsecond, time travel will contain two snapshots, and the test fails too. And so I make a sleep here…


The root cause is the time precision of the snapshot is nanosecond, but we can only use TIMESTAMP[0, 3, 6] in sql now.

Except for the test, I don’t think it’s a serious issue. In general, we won’t use the timestamp of a snapshot to make time travel, instead of snapshot id.

Reply:
Thanks for the informative description!